### PR TITLE
refactor(ir): require tmp scratch operand on tile.transpose

### LINKED ISSUE #1398: a2a3: VecTile [16,1] column load from GlobalTensor fails TLOAD layout check
## Summary

Loading a column-shaped Vec tile (`[16, 1]`) from a row-major/ND global tensor fails during a2a3 incore compilation with:

```text
TLOAD(VecTile, GlobalTensor) only support ND2ND/DN2DN/NZ2NZ
```

This blocks T-axis tiled kernels that need to load a per-row scale column and broadcast it across a wider tile, such as `hc_pre.mix_x` using `[T_TILE, D_CHUNK]` tiles.

## Minimal reproducer

Complete standalone reproducer:

```python
# Copyright (c) PyPTO Contributors.
# This program is free software, you can redistribute it and/or modify it under the terms and conditions of
# CANN Open Software License Agreement Version 2.0 (the "License").
# Please refer to the License for details. You may not use this file except in compliance with the License.
# THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
# INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
# See LICENSE in the root of the software repository for the full text of the License.
# -----------------------------------------------------------------------------------------------------------
"""Reproducer for a2a3 VecTile column-load lowering.

The kernel loads ``scale[:, 0:1]`` as a ``[16, 1]`` Vec tile from a global
``[16, 8]`` tensor, then broadcasts it across ``x`` with ``row_expand_mul``.
On a2a3 this currently fails during incore compilation with:

    TLOAD(VecTile, GlobalTensor) only support ND2ND/DN2DN/NZ2NZ
"""

import pypto.language as pl

ROWS = 16
COLS = 64
SCALE_COLS = 8


def build_repro_program():
    @pl.program
    class VecColumnLoadRepro:
        @pl.function(type=pl.FunctionType.Opaque)
        def repro(
            self,
            x: pl.Tensor[[ROWS, COLS], pl.FP32],
            scale: pl.Tensor[[ROWS, SCALE_COLS], pl.FP32],
            y: pl.Out[pl.Tensor[[ROWS, COLS], pl.FP32]],
        ) -> pl.Tensor[[ROWS, COLS], pl.FP32]:
            with pl.at(level=pl.Level.CORE_GROUP, name_hint="vec_col_load"):
                col = pl.load(
                    scale,
                    [0, 0],
                    [ROWS, 1],
                    target_memory=pl.MemorySpace.Vec,
                )
                x_tile = pl.load(
                    x,
                    [0, 0],
                    [ROWS, COLS],
                    target_memory=pl.MemorySpace.Vec,
                )
                y_tile = pl.row_expand_mul(x_tile, col)
                y = pl.store(y_tile, [0, 0], y)
            return y

    return VecColumnLoadRepro


def build_specs():
    import torch
    from golden import TensorSpec

    return [
        TensorSpec("x", [ROWS, COLS], torch.float32, init_value=torch.randn),
        TensorSpec("scale", [ROWS, SCALE_COLS], torch.float32, init_value=torch.randn),
        TensorSpec("y", [ROWS, COLS], torch.float32, is_output=True),
    ]


def golden(values):
    values["y"][:] = values["x"] * values["scale"][:, 0:1]


if __name__ == "__main__":
    import argparse

    from golden import RunConfig, run

    parser = argparse.ArgumentParser()
    parser.add_argument("-p", "--platform", type=str, default="a2a3",
                        choices=["a2a3", "a2a3sim", "a5", "a5sim"])
    parser.add_argument("-d", "--device", type=int, default=0)
    args = parser.parse_args()

    result = run(
        program=build_repro_program(),
        specs=build_specs(),
        golden_fn=golden,
        config=RunConfig(
            rtol=1e-5,
            atol=1e-5,
            compile=dict(dump_passes=True),
            runtime=dict(platform=args.platform, device_id=args.device),
        ),
    )
    if not result.passed:
        if result.error:
            print(result.error)
        raise SystemExit(1)
```

## Reproduction command

Run through the NPU task queue:

```bash
task-submit --device 11 \
  --env PYTHONPATH --env ASCEND_HOME_PATH --env PTOAS_ROOT --env PTO_ISA_ROOT \
  --max-time 120 \
  --run "cd /data/yangyaodong/code/pto_workspace/modules/pypto-lib && python examples/bugs/repro_vec_col_load.py -p a2a3 -d {}"
```

Task used while reproducing:

```text
task_20260518_192628_356422018233
```

## Observed result

The PyPTO compile stage completes, then runtime-side incore compilation fails:

```text
[RUN] compile done (0.60s)
[RUN] runtime ...
[Incore] Compilation failed:
.../TLoad.hpp:459:9: error: static assertion failed due to requirement 'isSameLayout': Fix: TLOAD(VecTile, GlobalTensor) only support ND2ND/DN2DN/NZ2NZ!
...
TLOAD_TILE_IMPL<pto::Tile<pto::TileType::Vec, float, 16, 1, pto::BLayout::ColMajor, ...>,
                 pto::GlobalTensor<float, pto::Shape<1, 1, 1, 16, 1>, pto::Stride<128, 128, 128, 8, 1>, pto::Layout::ND>>
...
TLOAD(v11, v15);
```

## Expected result

One of:

- PyPTO lowers `[16, 1]` Vec column loads from ND global tensors to a legal a2a3 load sequence, or
- PyPTO rejects this shape/layout earlier with an actionable diagnostic and suggests the supported workaround.

## Notes

A related `hc_pre.mix_x` optimization wants to load `pre_val[t0:t0+T_TILE, h:h+1]` and use it as a per-row scale for a `[T_TILE, D_CHUNK]` tile. Direct `[TILE,1]` loading hits this compile error. Loading a wider `[TILE, HC_PAD]` tile avoids this exact compile error but led to runtime timeout in the larger kernel, so a legal direct column-load lowering would be useful.

### LINKED ISSUE #1402: [Bug] Incore pl.transpose emits pto.alloc_tile with dynamic valid shape but no valid_row operand
### Component

Codegen

### Description

`pl.transpose` inside an incore `pl.at` scope lowers to invalid PTO MLIR on a2a3. The generated `ttrans_tmp` tile has dynamic `v_row=?` / `v_col=?`, but the `pto.alloc_tile` operation does not provide `valid_row` / `valid_col` operands, so ptoas rejects the file before C++ generation.

This blocks using incore tile transpose as a workaround for T-axis tiled kernels that want to physically transpose a small `[T_TILE, HC_PAD]` tile before row-broadcast use.

Diagnosis: **pypto codegen**. The failure happens before runtime execution: PyPTO emits invalid `.pto` for the temporary transpose buffer. ptoas reports a malformed `pto.alloc_tile` op.

Related but not duplicate:

- #1267: dynamic valid shape on `pto.subview`
- #1398: `[16,1]` Vec column load from ND GM layout check

### Steps to Reproduce

Save this as `examples/beginner/incore_transpose_demo.py` in `pypto-lib`:

```python
# Copyright (c) PyPTO Contributors.
# This program is free software, you can redistribute it and/or modify it under the terms and conditions of
# CANN Open Software License Agreement Version 2.0 (the "License").
# Please refer to the License for details. You may not use this file except in compliance with the License.
# THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
# INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
# See LICENSE in the root of the software repository for the full text of the License.
# -----------------------------------------------------------------------------------------------------------
"""Minimal demo for an incore tile transpose."""

import pypto.language as pl


ROWS = 16
COLS = 8


@pl.jit
def incore_transpose(
    x: pl.Tensor[[ROWS, COLS], pl.FP32],
    y: pl.Out[pl.Tensor[[COLS, ROWS], pl.FP32]],
):
    with pl.at(level=pl.Level.CORE_GROUP, name_hint="tile_transpose"):
        x_tile = pl.load(
            x,
            [0, 0],
            [ROWS, COLS],
            target_memory=pl.MemorySpace.Vec,
        )
        y_tile = pl.transpose(x_tile, axis1=0, axis2=1)
        y = pl.store(y_tile, [0, 0], y)
    return y


def build_tensor_specs():
    import torch
    from golden import TensorSpec

    return [
        TensorSpec("x", [ROWS, COLS], torch.float32, init_value=torch.randn),
        TensorSpec("y", [COLS, ROWS], torch.float32, is_output=True),
    ]


def golden_transpose(tensors):
    tensors["y"][:] = tensors["x"].transpose(0, 1).contiguous()


if __name__ == "__main__":
    import argparse
    from golden import RunConfig, run_jit

    parser = argparse.ArgumentParser()
    parser.add_argument(
        "-p",
        "--platform",
        type=str,
        default="a2a3",
        choices=["a2a3", "a2a3sim", "a5", "a5sim"],
    )
    parser.add_argument("-d", "--device", type=int, default=0)
    parser.add_argument("--enable-l2-swimlane", action="store_true", default=False)
    args = parser.parse_args()

    result = run_jit(
        fn=incore_transpose,
        specs=build_tensor_specs(),
        golden_fn=golden_transpose,
        config=RunConfig(
            rtol=1e-5,
            atol=1e-5,
            compile=dict(dump_passes=True),
            runtime=dict(
                platform=args.platform,
                device_id=args.device,
                enable_l2_swimlane=args.enable_l2_swimlane,
            ),
        ),
    )
    if not result.passed:
        if result.error:
            print(result.error)
        raise SystemExit(1)
```

Run with a real NPU through the task queue:

```bash
task-submit --device auto \
  --env PYTHONPATH --env ASCEND_HOME_PATH --env PTOAS_ROOT --env PTO_ISA_ROOT \
  --max-time 300 \
  --run "cd /data/yangyaodong/code/pto_workspace/modules/pypto-lib && python examples/beginner/incore_transpose_demo.py -p a2a3 -d {}"
```

The reproducing task was:

```text
task_20260519_110525_215122727554
```

### Expected Behavior

The incore transpose kernel should compile and run. The generated `.pto` should allocate any temporary tile used by `pto.ttrans` with valid shape operands when its result type has dynamic `v_row` / `v_col`, or use a static valid shape when possible for the `[16, 8] -> [8, 16]` case.

### Actual Behavior

Compilation fails in ptoas:

```text
Failed to compile function 'tile_transpose': ptoas compilation failed:
loc("build_output/_jit_incore_transpose_20260519_110527/ptoas/tile_transpose.pto":10:17):
error: 'pto.alloc_tile' op valid_row operand is required because result type v_row is ?
Error: Failed to parse MLIR.
```

The offending generated IR is:

```mlir
module attributes {pto.target_arch = "a2a3"} {
  func.func @tile_transpose(%arg0: !pto.ptr<f32>, %arg1: !pto.ptr<f32>) attributes {pto.kernel_kind = #pto.kernel_kind<vector>} {
  %c0_i64 = arith.constant 0 : i64
  %c16_index = arith.constant 16 : index
  %c8_index = arith.constant 8 : index
  %c1_index = arith.constant 1 : index
  %c0_index = arith.constant 0 : index
  %x__ssa_v0_view = pto.make_tensor_view %arg0, shape = [%c16_index, %c8_index], strides = [%c8_index, %c1_index] {layout = #pto.layout<nd>}: !pto.tensor_view<?x?xf32>
  %y__ssa_v0_view = pto.make_tensor_view %arg1, shape = [%c8_index, %c16_index], strides = [%c16_index, %c1_index] {layout = #pto.layout<nd>}: !pto.tensor_view<?x?xf32>
  %ttrans_tmp = pto.alloc_tile : !pto.tile_buf<loc=vec, dtype=f32, rows=16, cols=8, v_row=?, v_col=?, blayout=row_major, slayout=none_box, fractal=512, pad=0>
  %x_tile__ssa_v0 = pto.alloc_tile addr = %c0_i64 valid_row = %c16_index valid_col = %c8_index : !pto.tile_buf<loc=vec, dtype=f32, rows=16, cols=8, v_row=?, v_col=?, blayout=row_major, slayout=none_box, fractal=512, pad=0>
  %x__ssa_v0_pview = pto.partition_view %x__ssa_v0_view, offsets = [%c0_index, %c0_index], sizes = [%c16_index, %c8_index] : !pto.tensor_view<?x?xf32> -> !pto.partition_tensor_view<16x8xf32>
  pto.tload ins(%x__ssa_v0_pview : !pto.partition_tensor_view<16x8xf32>) outs(%x_tile__ssa_v0 : !pto.tile_buf<loc=vec, dtype=f32, rows=16, cols=8, v_row=?, v_col=?, blayout=row_major, slayout=none_box, fractal=512, pad=0>)
  %y_tile__ssa_v0 = pto.alloc_tile addr = %c0_i64 valid_row = %c8_index valid_col = %c16_index : !pto.tile_buf<loc=vec, dtype=f32, rows=8, cols=16, v_row=?, v_col=?, blayout=row_major, slayout=none_box, fractal=512, pad=0>
  pto.ttrans ins(%x_tile__ssa_v0, %ttrans_tmp : !pto.tile_buf<loc=vec, dtype=f32, rows=16, cols=8, v_row=?, v_col=?, blayout=row_major, slayout=none_box, fractal=512, pad=0>, !pto.tile_buf<loc=vec, dtype=f32, rows=16, cols=8, v_row=?, v_col=?, blayout=row_major, slayout=none_box, fractal=512, pad=0>) outs(%y_tile__ssa_v0 : !pto.tile_buf<loc=vec, dtype=f32, rows=8, cols=16, v_row=?, v_col=?, blayout=row_major, slayout=none_box, fractal=512, pad=0>)
  %y__ssa_v0_pview = pto.partition_view %y__ssa_v0_view, offsets = [%c0_index, %c0_index], sizes = [%c8_index, %c16_index] : !pto.tensor_view<?x?xf32> -> !pto.partition_tensor_view<8x16xf32>
  pto.tstore ins(%y_tile__ssa_v0 : !pto.tile_buf<loc=vec, dtype=f32, rows=8, cols=16, v_row=?, v_col=?, blayout=row_major, slayout=none_box, fractal=512, pad=0>) outs(%y__ssa_v0_pview : !pto.partition_tensor_view<8x16xf32>)
  return
  }
}
```

### Git Commit ID

41b9ce52b15abd5aa456747c5b268866b15814a3

### NPU Kind

Ascend 910B

### Host Platform

Linux (aarch64)

### Additional Context

Environment used while reproducing:

| Component | Version |
|---|---|
| pypto-lib | `b06a6eb` |
| pypto | `41b9ce52` (branch: `main`) |
| simpler | `a94d5140` |
| ptoas | `0.40` |
| CANN | `not detected` |
| Python | `3.10.20` |

The generated PTO file was at:

```text
build_output/_jit_incore_transpose_20260519_110527/ptoas/tile_transpose.pto
```

The codegen error file was at:

```text
build_output/_jit_incore_transpose_20260519_110527/report/codegen_errors.txt
```

### PR DESCRIPTION
**Pull request by YunjiQin**
## Summary

- `tile.transpose` now takes `(input, axis1, axis2, tmp)` — `tmp` is a scratch buffer (same shape/dtype as input) required by `pto.ttrans`. Pre-allocating it as a separate `tile.create` lets the memory allocator assign a real UB address before backend codegen.
- This is mandatory at `--pto-level=level3`, where the previous codegen-internal `AllocNewTileBuf` fallback produced unaddressed tiles for `ForStmt`-result or `tile.reshape`-view sources.
- Python IR-bindings and the language DSL expose a trailing optional `tmp` / `tmp_tile` parameter that auto-emits `tile.create` when omitted, so user-facing 3-arg calls still work and printer-emitted 4-arg calls round-trip through the existing parser without changes.

## Layers touched

- **IR op**: `DeduceTileTransposeType` + `REGISTER_OP` enforce the 4-arg form
- **Codegen**: `pto.ttrans` simplified to the single 4-arg path
- **Pass**: `ConvertTensorToTileOps` lowers `tensor.transpose` to `tile.create` + 4-arg `tile.transpose`
- **Pass**: `FlattenTileNdTo2D` emits the tmp tile when wrapping `batch_matmul` operands; wrapper detection updated to the new arg order
- **Python/DSL**: bindings + language surface the optional trailing `tmp` parameter

## Testing

- [x] Added `ColumnLoadRowExpandMulCase` st regression exercising the `ForStmt`-result transpose path that previously failed at level3
- [x] Added unit tests for the new op arity and for the `ConvertTensorToTileOps` lowering
- [x] Existing tile/transpose tests updated to the new arg order

## Related Issues

Fixes #1398
Fixes #1402

### CODE REVIEW

File: src/ir/op/tile_ops/transform.cpp [hunk 1]
```diff
@@ -317,25 +317,33 @@ TypePtr DeduceTileReshapeType(const std::vector<ExprPtr>& args,
...
+  CHECK(tmp_type) << "tile.transpose: fourth argument (tmp) must be TileType, but got "
+                  << args[3]->GetType()->TypeName();
 
-  const auto& input_shape = tile_type->shape_;
+  CHECK(input_type->dtype_ == tmp_type->dtype_)
+      << "tile.transpose: tmp dtype must match input dtype, got input=" << input_type->dtype_.ToString()
+      << " tmp=" << tmp_type->dtype_.ToString();
+  CHECK(input_type->shape_.size() == tmp_type->shape_.size())
+      << "tile.transpose: tmp rank must match input rank, got input rank=" << input_type->shape_.size()
+      << " tmp rank=" << tmp_type->shape_.size();
```
[the remark this answers is not in the document]
Comment (YunjiQin) [reply src/ir/op/tile_ops/transform.cpp:337]: Skipping — at the ISA/`pto.ttrans` level, `tmp` only needs to match `input` in rank (and dtype), not in per-dim shape. The current rank+dtype CHECKs already capture the binding contract; stricter per-dim equality would reject legal callers that pre-allocate a compatible scratch with differently-sized but rank-matching dims.

**Review by lyfne123 — APPROVED** [review]
